rust_src/src/fns.rs

```diff
@@ -335,6 +335,50 @@ pub fn reverse(seq: LispObject) -> LispObject {
     }
 }
 
+/// Reverse order of items in a list, vector or string SEQ.
+/// If SEQ is a list, it should be nil-terminated.
+/// This function may destructively modify SEQ to produce the value.
+#[lisp_fn]
+pub fn nreverse(mut seq: LispObject) -> LispObject {
+    if seq.is_nil() {
+        return seq;
+    } else if seq.is_string() {
+        return reverse(seq);
+    } else if let Some(cons) = seq.as_cons() {
+        let mut iter = cons
+            .iter_tails(LispConsEndChecks::on, LispConsCircularChecks::on)
+            .peekable();
+        let mut prev = Qnil;
+        while let Some(tail) = iter.next() {
+            // Invoked to advance the underlying iterator
+            iter.peek();
```

Comment:
From the Rust docs (https://doc.rust-lang.org/std/iter/struct.Peekable.html):

> Returns a reference to the next() value without advancing the iterator.

That call to `peek()` is not doing anything.

Reply:
This contradicts the documentation in [Iterator::peekable](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.peekable), which states:

> Note that the underlying iterator is still advanced when peek is called for the first time: In order to retrieve the next element, next is called on the underlying iterator, hence any side effects (i.e. anything other than fetching the next value) of the next method will occur.

And from looking at how the code actually behaves, peek does invoke side effects, as the iterator is correctly advanced before the cdr is set.

Reply:
Looking into it further, I think the documentation for the peek method is poorly worded, I think when it says "without advancing the Iterator", it's referring to the wrapping iterator, and not the underlying one.

Reply:
Weird. I would not expect `peek` to change the iterator. What a horrible name.

Reply:
Wait.

```
let xs = [1, 2, 3];

let mut iter = xs.iter().peekable();

// peek() lets us see into the future
assert_eq!(iter.peek(), Some(&&1));
assert_eq!(iter.peek(), Some(&&1));
```

How can `peek` be advancing the iterator if the same value is there?? The examples on the site match my intuition. Do we have a bug in how our iterator works?

Reply:
No, it works as intended. peek() only advances the underlying iterator the first time it is invoked after a call to next()

Reply:
Does this playground match your expectation? https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7c0334c9597396569f02bbd9d3d50e10

Reply:
My understanding is this loop:
```
let mut it = thing.iter();
loop {
    match it.next() {
        Some(value) => { some action on value },
        None => break,
    }
}
```
and this loop:
```
let mut it = thing.iter().peekable();
loop {
    match it.peek() {
        Some(p) => { println!("Peek work: {}", p); },
        None => { println!("Nothing else to see."); },
    };
    match it.next() {
        Some(value) => { some action on value },
        None => break,
    };
}
```

Should yield the same values and essentially be identical.

Reply:
Yes, the playground matches my expectation. And the reason that invoking peek() changes behaviour is that it advances the underlying iterator before the cdr is changed, effectively caching which next value to use.

If peek() is not invoked, every time next() is called it would advance the iterator to the new cdr, which is the cons cell the used to be the car of the current one.

The caching behaviour mirrors how the C code defines a `next = XCDR(tail)` value at the start of every loop, and invokes `tail = next` at the end of the loop (or rather, the last part of the for loop. but it's functionally the same).